AudibleApi/Cryptography/PrivateKey.cs

```diff
@@ -52,51 +50,18 @@ private static RSACryptoServiceProvider CreateRsaProviderFromPrivateKey(string p
                 .Trim();
 
             var asn1EncodedPrivateKey = Convert.FromBase64String(privateKey);
-            var keySequence = Asn1Value.Parse(asn1EncodedPrivateKey);
-
-            if (keySequence.Children.Count >= 3 && keySequence.Children[2].Type == Asn1Tag.PrimitiveOctetString)
-                keySequence = Asn1Value.Parse(keySequence.Children[2].Value);
-
             var RSA = new RSACryptoServiceProvider();
-            var RSAparams = new RSAParameters();
-
-            //Guess the key size based on modulus size. Assume key size is multiple of 32 bits.
-            int keySizeGuess = (int)Math.Round(keySequence.Children[1].Value.Length / 4d, 0) * 32;
-
-            RSAparams.Modulus = ValidateInt(keySequence.Children[1].Value, keySizeGuess, 8);
-            RSAparams.Exponent = keySequence.Children[2].Value;
-            RSAparams.D = ValidateInt(keySequence.Children[3].Value, keySizeGuess, 8);
-            RSAparams.P = ValidateInt(keySequence.Children[4].Value, keySizeGuess, 16);
-            RSAparams.Q = ValidateInt(keySequence.Children[5].Value, keySizeGuess, 16);
-            RSAparams.DP = ValidateInt(keySequence.Children[6].Value, keySizeGuess, 16);
-            RSAparams.DQ = ValidateInt(keySequence.Children[7].Value, keySizeGuess, 16);
-            RSAparams.InverseQ = ValidateInt(keySequence.Children[8].Value, keySizeGuess, 16);
-
-            RSA.ImportParameters(RSAparams);
-            return RSA;
-        }
-
-        /// <summary>
-        /// Trim or add leading zeroes to ensure <see cref="RSAParameters"/> are valid.
-        /// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-wcce/5cf2e6b9-3195-4f85-bc18-05b50e6d4e11?redirectedfrom=MSDN
-        /// </summary>
-        private static byte[] ValidateInt(byte[] integer, int keySize, int quotient)
-        {
-            var expectedSize = keySize / quotient;
-
-            if (integer.Length == expectedSize)
-                return integer;
-            else if (integer.Length == expectedSize + 1 && integer[0] == 0)
-                return integer[1..];
-            else if (integer.Length == expectedSize - 1)
+            try
             {
-                var bts = new byte[expectedSize];
-                Buffer.BlockCopy(integer, 0, bts, 1, integer.Length);
-                return bts;
+				//Try importing the iPhone private key, which is PKCS #1.
+				RSA.ImportRSAPrivateKey(asn1EncodedPrivateKey, out _);
             }
-            else
-                //If this happens, it won't happen for all key parameters, so logging the real value will not reveal the complete key.
-                throw new CryptographicException($"Unable to parse rsa parameter integer: {BitConverter.ToString(integer)}");
+            catch(CryptographicException)
+            {
+				//Fallback to importing the Android private key, which is PKCS#8.
+				RSA.ImportPkcs8PrivateKey(asn1EncodedPrivateKey, out _);
+            }
+            return RSA;
```

Comment:
Now that I understand _what_ these private key blobs are, parse them using dotnet's built-in parser.

Reply:
Holy cow -- this is all built into the base libraries now? What I wouldn't have given for that 5ish years ago